Ensure completion handler from WebsiteDataStoreClient::getScreenTimeURLs runs on main thread

jesxilin opened the pull request:
#### b90ad95246b984c810ce341e8827c00075c2d2bd
<pre>
Ensure completion handler from WebsiteDataStoreClient::getScreenTimeURLs runs on main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=289021">https://bugs.webkit.org/show_bug.cgi?id=289021</a>
<a href="https://rdar.apple.com/146061712">rdar://146061712</a>

Reviewed by Aditya Keerthi.

Ensure we are on main thread when calling captured completion handler.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:

Canonical link: <a href="https://commits.webkit.org/291715@main">https://commits.webkit.org/291715@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77df07a896de8b420cd04e27bc19210d5ded7aeb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93798 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/13372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3111 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/98804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/44324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21814 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/98804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/44324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96800 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/98804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/43639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/2464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/100839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/20850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/100839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/21102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/80710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/100839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15042 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/20834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/20521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/23981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/22262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -189,18 +189,20 @@ void getScreenTimeURLs(std::optional<WTF::UUID> identifier, CompletionHandler<vo
-                    result.add(url);
-            }
-            completionHandler(WTFMove(result));
+            ensureOnMainRunLoop([completionHandler = WTFMove(completionHandler), urls = WTFMove(urls), error = retainPtr(error)] () mutable {
```
C++23
```suggestion
            ensureOnMainRunLoop([completionHandler = WTFMove(completionHandler), urls = WTFMove(urls), error = retainPtr(error)] mutable {
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -189,18 +189,20 @@ void getScreenTimeURLs(std::optional<WTF::UUID> identifier, CompletionHandler<vo
-                    result.add(url);
-            }
-            completionHandler(WTFMove(result));
+            ensureOnMainRunLoop([completionHandler = WTFMove(completionHandler), urls = WTFMove(urls), error = retainPtr(error)] () mutable {
```
You need to retain the URLs.

aprotyas replied:
yes

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -189,18 +189,20 @@ void getScreenTimeURLs(std::optional<WTF::UUID> identifier, CompletionHandler<vo
-            }
-            completionHandler(WTFMove(result));
+            ensureOnMainRunLoop([completionHandler = WTFMove(completionHandler), urls = WTFMove(urls), error = retainPtr(error)] () mutable {
+                if (error.get() != nil) {
```
```suggestion
                if (error) {
```

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -189,18 +189,20 @@ void getScreenTimeURLs(std::optional<WTF::UUID> identifier, CompletionHandler<vo
+                for (NSURL *site in urls.get()) {
+                    URL url { site };
+                    if (url.isValid())
+                        result.add(url);
```
`WTFMove(url)` would be slightly more efficient here.

jesxilin committed:
Ensure completion handler from WebsiteDataStoreClient::getScreenTimeURLs runs on main thread
https://bugs.webkit.org/show_bug.cgi?id=289021
rdar://146061712

Reviewed by Aditya Keerthi.

Ensure we are on main thread when calling captured completion handler.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:

Canonical link: https://commits.webkit.org/291715@main
